lib/readelf.c

```diff
@@ -153,30 +163,46 @@ static bool _is_elf_type(const char *path, int type)
  * Return true if a specified file is an ELF shared library file, that
  * is, of type ET_DYN.
  */
-bool is_elf_shared_library(const char *path)
+bool is_elf_shared_library(rpmfile_entry_t *file)
 {
-    return _is_elf_type(path, ET_DYN);
+    /* -1 "no", 1 "yes", 0 "don't know yet" */
+    if (file->is_elf_shared_library != 0)
+        return (file->is_elf_shared_library + 1) >> 1; /* return 1/0 for "yes" / "no" */
+
+    bool r = _is_elf_type(file, ET_DYN);
+    file->is_elf_shared_library = r ? 1 : -1;
+    return r;
 }
 
 /*
  * Return true if a specified file is an ELF executable file, that
  * is, of type ET_EXEC.
  */
-bool is_elf_executable(const char *path)
+bool is_elf_executable(rpmfile_entry_t *file)
 {
-    return _is_elf_type(path, ET_EXEC);
+    /* -1 "no", 1 "yes", 0 "don't know yet" */
+    if (file->is_elf_executable != 0)
+        return (file->is_elf_executable + 1) >> 1; /* return 1/0 for "yes" / "no" */
+
+    bool r = _is_elf_type(file, ET_EXEC);
+    file->is_elf_executable = r ? 1 : -1;
+    return r;
 }
 
 /*
  * Return true for any ELF file.
  */
-bool is_elf_file(const char *path)
+bool is_elf_file(rpmfile_entry_t *file)
 {
+    /* -1 "no", 1 "yes", 0 "don't know yet" */
+    if (file->is_elf_file != 0)
```

Comment:
Please use { } around the block of this if statement even if it is only a single line.

Reply:
done